sync/pooled_worker_pool.go

```diff
@@ -0,0 +1,122 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package sync implements synchronization facililites such as worker pools.
+package sync
+
+import "math/rand"
+
+const (
+	defaultNumShards = 10
+)
+
+var (
+	// Var for testing purposes
+	defaultKillWorkerProbability = 0.0001
+)
+
+// PooledWorkerPool provides a pool for goroutines, but unlike WorkerPool,
+// the actual goroutines themselves are re-used. This can be useful from a
+// performance perspective in scenarios where the allocation and growth of
+// the new goroutine and its stack is a bottleneck. Specifically, if the
+// work function being performed has a very deep call-stack, calls to
+// runtime.morestack can dominate the workload. Re-using existing goroutines
+// allows the stack to be grown once, and then re-used for many invocations.
+//
+// In order to prevent abnormally large goroutine stacks from persisting over
+// the life-cycle of an application, the PooledWorkerPool will randomly kill
+// existing goroutines and spawn a new one.
+//
+// The PooledWorkerPool also implements sharding of its underlying worker channels
+// to prevent excessive lock contention.
+type PooledWorkerPool interface {
+	// Init initializes the pool.
+	Init()
+
+	// Go waits until the next worker becomes available and executes it.
+	Go(work Work)
+}
+
+type pooledWorkerPool struct {
+	workChs   []chan workerInstruction
+	numShards int
+}
+
+type workerInstruction struct {
+	work      Work
+	shouldDie bool
+}
+
+// NewPooledWorkerPool creates a new worker pool.
+func NewPooledWorkerPool(size int) PooledWorkerPool {
+	numShards := defaultNumShards
+	if size < numShards {
+		numShards = size
+	}
+
+	workChs := make([]chan workerInstruction, numShards)
+	for i := range workChs {
+		workChs[i] = make(chan workerInstruction, size/numShards)
+	}
+
+	return &pooledWorkerPool{
+		workChs:   workChs,
+		numShards: numShards,
+	}
+}
+
+func (p *pooledWorkerPool) Init() {
+	for _, workCh := range p.workChs {
+		for i := 0; i < cap(workCh); i++ {
+			p.spawnWorker(workCh)
+		}
+	}
+}
+
+func (p *pooledWorkerPool) Go(work Work) {
+	instruction := workerInstruction{work: work}
+	randVal := rand.Float64()
+	killWorker := randVal < defaultKillWorkerProbability
+	instruction.shouldDie = killWorker
+
+	// Pick an index based on the random value. Example:
+	// 		numShards = 10
+	// 		randVal = 0.14
+	// 		int(0.14 * float64(10)) = 1
+	// 		1 % 10 == 1
+	workChIdx := int((randVal * float64(p.numShards))) % p.numShards
```

Comment:
Nit: Can you benchmark all this type casting VS generating a random int to use here. You can generate a random int to use here along with probability, just multiply your probability % option with the max int range you're going to use and check against that int.

Reply:
I did some micro-benchmarks and there doesn't seem to be any (reliable) performance difference between getting a random float, and gettting a random float and then casting it to an int. We're talking about 20ns operations here so I don't think this will really impact anything compared to the amount of "work" being done. Plus I think its a little easier to follow this way.